tools/database/src/analyse_gas_usage.rs

```diff
@@ -0,0 +1,348 @@
+        let shard_uid: ShardUId = ShardUId::from_shard_id_and_layout(shard_id, &shard_layout);
+
+        let mut gas_usage_in_shard = GasUsageInShard::new();
+
+        // The outcome of each transaction and receipt executed in this chunk is saved in the database as an ExecutionOutcome.
+        // Go through all ExecutionOutcomes from this chunk and record the gas usage.
+        let outcome_ids: Vec<CryptoHash> =
+            chain_store.get_outcomes_by_block_hash_and_shard_id(block.hash(), shard_id).unwrap();
+        for outcome_id in outcome_ids {
+            let outcome: ExecutionOutcome = chain_store
+                .get_outcome_by_id_and_block_hash(&outcome_id, block.hash())
+                .unwrap()
+                .unwrap()
+                .outcome;
+
+            // Sanity check - make sure that the executor of this outcome belongs to this shard
+            let account_shard: ShardId =
+                account_id_to_shard_id(&outcome.executor_id, &shard_layout);
+            assert_eq!(account_shard, shard_id);
+
+            gas_usage_in_shard.add_used_gas(outcome.executor_id, outcome.gas_burnt.into());
+        }
+
+        result.add_gas_usage_in_shard(shard_uid, gas_usage_in_shard);
+    }
+
+    result
+}
+
+/// A struct that can be used to find N biggest accounts by gas usage in an efficient manner.
+struct BiggestAccountsFinder {
+    accounts: BTreeSet<(BigGas, AccountId)>,
+    accounts_num: usize,
+}
+
+impl BiggestAccountsFinder {
+    pub fn new(accounts_num: usize) -> BiggestAccountsFinder {
+        BiggestAccountsFinder { accounts: BTreeSet::new(), accounts_num }
+    }
+
+    pub fn add_account_stats(&mut self, account: AccountId, used_gas: BigGas) {
+        self.accounts.insert((used_gas, account));
+
+        // If there are more accounts than desired, remove the one with the smallest gas usage
+        while self.accounts.len() > self.accounts_num {
+            self.accounts.pop_first();
+        }
+    }
+
+    pub fn get_biggest_accounts(&self) -> impl Iterator<Item = (AccountId, BigGas)> + '_ {
+        self.accounts.iter().rev().map(|(gas, account)| (account.clone(), *gas))
+    }
+}
+
+fn analyse_gas_usage(
+    blocks_iter: impl Iterator<Item = Block>,
+    chain_store: &ChainStore,
+    epoch_manager: &EpochManager,
+) {
+    // Gather statistics about gas usage in all of the blocks
+    let mut blocks_count: usize = 0;
+    let mut first_analysed_block: Option<(BlockHeight, CryptoHash)> = None;
+    let mut last_analysed_block: Option<(BlockHeight, CryptoHash)> = None;
+
+    let mut gas_usage_stats = GasUsageStats::new();
+
+    for block in blocks_iter {
+        blocks_count += 1;
+        if first_analysed_block.is_none() {
+            first_analysed_block = Some((block.header().height(), *block.hash()));
+        }
+        last_analysed_block = Some((block.header().height(), *block.hash()));
+
+        let gas_usage_in_block: GasUsageStats =
```

Comment:
I personally like to have as many type annotations as possible. In an IDE this doesn't matter because it adds annotations itself, but there are places where this isn't the case - e.g when reviewing the code on github. I feel like having them makes it easier to figure out what's going on - here I would instantly know that we are merging instances of `GasUsageStats`.

Reply:
All right, but it does make editing code a bit harder - because then you need to adjust the types on every step as well. Generally speaking the current convention in nearcore is to only add type annotation when needed and we should stick to that.

Reply:
Okay if that's the convention then I will adjust :+1:. I don't really like it, but OK. 

I dislike reading code without type annotations, reading Python is always a pain. One could argue that code is read more often than it's written, so it's more important to make it as readable as possible.

Reply:
Sounds like a good topic for zulip or Monday core team discussion :)